File: DPPP/DummyStep.cc
// GainCal.cc: DPPP step class to DummyStep visibilities
// Copyright (C) 2020 ASTRON (Netherlands Institute for Radio Astronomy)
// SPDX-License-Identifier: GPL-3.0-or-later
//
// @author Tammo Jan Dijkema

#include "DummyStep.h"

#include <iostream>

#include "../Common/ParameterSet.h"
#include "../Common/Timer.h"

#include <stddef.h>
#include <string>
#include <sstream>
#include <utility>
#include <vector>

using namespace casacore;

namespace DP3 {
namespace DPPP {

DummyStep::DummyStep(DPInput* input, const ParameterSet& parset,
                     const string& prefix)
    : itsInput(input) {}

DummyStep::~DummyStep() {}

void DummyStep::updateInfo(const DPInfo& infoIn) {
  info() = infoIn;
  info().setNeedVisData();
  info().setWriteData();
}

void DummyStep::show(std::ostream& os) const {
  os << "DummyStep " << itsName << endl;
}

void DummyStep::showTimings(std::ostream& os, double duration) const {
  os << "  ";
  FlagCounter::showPerc1(os, itsTimer.getElapsed(), duration);
  os << " DummyStep " << itsName << endl;
}

bool DummyStep::process(const DPBuffer& bufin) {
  itsTimer.start();
  itsBuffer.copy(bufin);
  itsInput->fetchUVW(bufin, itsBuffer, itsTimer);
  itsInput->fetchWeights(bufin, itsBuffer, itsTimer);

  itsTimer.stop();
  getNextStep()->process(itsBuffer);
  return false;
}

void DummyStep::finish() {
  // Let the next steps finish.
  getNextStep()->finish();
}
}  // namespace DPPP
}  // namespace DP3

File: PythonDPPP/dummystep.py
# Copyright (C) 2020 ASTRON (Netherlands Institute for Radio Astronomy)
# SPDX-License-Identifier: GPL-3.0-or-later

import dppp
import numpy as np

class DummyDPStep(dppp.DPStep):

    def __init__(self,parset, prefix):
        super().__init__()
        self.parset = parset
        self.dpbuffers = []
        self.prefix = prefix

    def show(self) :
        print()
        print("DummyDPStep")
        v = self.parset.getString(self.prefix + "somekey")
        print("    somekey: {v}")

    def process(self, dpbuffer) :

        # Accumulate buffers
        self.dpbuffers.append(dpbuffer)

        # If we have accumulated enough data, process it
        if len(self.dpbuffers) == 10:
            self.process_buffers()

            # Send processed data to the next step
            for dpbuffer in self.dpbuffers:
                self.process_next_step(dpbuffer)

            # Clear accumulated data
            self.dpbuffers = []

    def finish(self):

        # If there is any remaining data, process it
        if len(self.dpbuffers):
            self.process_buffers()
            for dpbuffer in self.dpbuffers:
                self.process_next_step(dpbuffer)
            self.dpbuffers = []

    def process_buffers(self):
        print()
        print(f"Processing a chunk of {len(self.dpbuffers)}.")

        # Enumerate accumulated data and display just the shapes
        for i, dpbuffer in enumerate(self.dpbuffers):
            data = np.array(dpbuffer.get_data(), copy=False)
            flags = np.array(dpbuffer.get_flags(), copy=False)
            uvw = np.array(dpbuffer.get_uvw(), copy=False)
            print(f"{i:4d} shape of data: ", data.shape)
            print(f"     shape of flags: ", flags.shape)
            print(f"     shape uvw: ", uvw.shape)

    def update_info(self, dpinfo) :
        super().update_info(dpinfo)
        print("DummyDPStep.update_info")
        self.info().set_need_vis_data()
        self.fetch_uvw = True
        self.fetch_weigths = True
